experimental/parser/lexer.go

```diff
@@ -0,0 +1,675 @@
+	}
+
+	// If the error is ErrRange we don't care, it will clamp to Infinity in
+	// the way we want, as noted by the comment above.
+	token.SetValue(tok, value)
+	return tok
+}
+
+// LexString lexes a string starting at the current cursor.
+//
+// The cursor position should be just before the string's first quote character.
+func (l *lexer) LexString() token.Token {
+	start := l.cursor
+	q := l.Pop()
+
+	// Seek to the end of the string, unescaping as we go. We do not
+	// materialize an unescaped string if this string does not require escaping.
+	var buf strings.Builder
+	var haveEsc bool
+escapeLoop:
+	for !l.Done() {
+		r := l.Pop()
+		if r == q {
+			break
+		}
+
+		// Warn if the user has a non-printable character in their string that isn't
+		// ASCII whitespace.
+		if !unicode.IsGraphic(r) && !strings.ContainsRune(" \n\t\r", r) {
+			l.Warnf("non-printable character in string literal").With(
+				report.Snippetf(l.SpanFrom(l.cursor-utf8.RuneLen(r)), "this is the rune U+%04x", r),
+			)
+		}
+
+		if r != '\\' {
+			// We intentionally do not legalize against literal \0 and \n. The above warning
+			// covers \0 and legalizing against \n is user-hostile. This is valuable for
+			// e.g. strings that contain CEL code.
+			//
+			// In other words, this limitation helps no one, so we ignore it.
+			if haveEsc {
+				buf.WriteRune(r)
+			}
+			continue
+		}
+
+		if !haveEsc {
+			buf.WriteString(l.Text()[start+1 : l.cursor-1])
+			haveEsc = true
+		}
+
+		r = l.Pop()
+		switch r {
+		// These are all the simple escapes.
+		case 'a':
+			buf.WriteByte('\a') // U+0007
+		case 'b':
+			buf.WriteByte('\b') // U+0008
+		case 'f':
+			buf.WriteByte('\f') // U+000C
+		case 'n':
+			buf.WriteByte('\n')
+		case 'r':
+			buf.WriteByte('\r')
+		case 't':
+			buf.WriteByte('\t')
+		case 'v':
+			buf.WriteByte('\v') // U+000B
+		case '\\', '\'', '"', '?':
+			buf.WriteRune(r)
+
+		// Octal escape. Need to eat the next twos rune if they're octal.
+		case '0', '1', '2', '3', '4', '5', '6', '7':
+			value := byte(r) - '0'
+			for i := 0; i < 2; i++ {
+				if l.Done() {
+					break escapeLoop
+				}
+				r = l.Peek()
+
+				// Check before consuming the rune. If we see e.g.
+				// an 8, we don't want to consume it.
+				if r < '0' || r > '7' {
+					break
+				}
+				_ = l.Pop()
+
+				value *= 8
+				value |= byte(r) - '0'
+			}
+			buf.WriteByte(value)
+
+		// Hex escapes.
+		case 'x', 'u', 'U':
+			var value uint32
+			var digits, consumed int
+			switch r {
+			case 'x':
+				digits = 2
+			case 'u':
+				digits = 4
+			case 'U':
+				digits = 8
+			}
+
+		digits:
+			for i := 0; i < digits; i++ {
+				if l.Done() {
+					break escapeLoop
+				}
+				r = l.Peek()
+
+				value *= 16
+				switch {
+				case r >= '0' && r <= '9':
+					value |= uint32(r) - '0'
+				case r >= 'a' && r <= 'f':
+					value |= uint32(r) - 'a' + 10
+				case r >= 'A' && r <= 'F':
+					value |= uint32(r) - 'A' + 10
+				default:
+					break digits
```

Comment:
This is not correct for `\u` and `\U` escapes, which must be exactly four or eight. Only `\x` can be short.

Reply:
Their length is validated elsewhere.